### PR TITLE
Create public api for dataset.

### PR DESCRIPTION
**Pull request by jenshnielsen**
* Move dond module to dataset where it belongs more naturally
* Let utils be a module that is independent of the rest of qcodes to avoid cirular dependenceis
     1. Move Slack and schema installation to a new module extensions
     2. Refactor a few other modules to import qcodes inside the functions. These will eventually be moved elsewhere
*  Import and document all methods used in dataset examples at the level of qcodes.dataset. 
* Add shims for old modules. These will not currently raise warnings. I suggest we turn these on after the new api is in a release to avoid confusion

This does not in general update examples to use the new imports. I suggest we do that after this pr has been merged. Can be done in small batches making review easier.

### CODE REVIEW

**Review by astafan8 — APPROVED** [review]

**Comment by jenshnielsen** [comment]
Should probably rename the dond module to clarify the difference between the dond function and module

File: qcodes/utils/dataset/doNd.py [hunk 1]
```diff
@@ -1,1054 +1,23 @@
...
-    for i, ax in enumerate(axes):
-        if save_pdf:
-            full_path = os.path.join(pdf_dif, f"{dataid}_{i}.pdf")
-            ax.figure.savefig(full_path, dpi=500)
-        if save_png:
-            full_path = os.path.join(png_dir, f"{dataid}_{i}.png")
-            ax.figure.savefig(full_path, dpi=500)
-    res = data, axes, cbs
-    return res
+# from qcodes.dataset.doNd import (
```
Comment (astafan8) [inline qcodes/utils/dataset/doNd.py:1]: i agree we should keep these imports working, but we can add a deprecation warning that triggers whenever an import from here is made.

**Review by astafan8 — COMMENTED** [review]
so happy to see doNd's being moved from utils to qcodes.dataset :)

File: qcodes/utils/threading.py [hunk 2]
```diff
@@ -138,12 +138,12 @@ def __repr__(self) -> str:
 
 def _instrument_to_param(
         params: Sequence[ParamMeasT]
-) -> Dict[Optional[str], Tuple[_BaseParameter, ...]]:
-
```
Comment (jenshnielsen) [inline qcodes/utils/threading.py:142]: This still internally imports parameters. Should it live somewhere else?

File: qcodes/utils/dataset/doNd.py (hunk 1 above)
Comment (jenshnielsen) [reply qcodes/utils/dataset/doNd.py]: Will be added after this is released. See comment above

**Comment by jenshnielsen** [comment]
@astafan8 Could you have a look again. The description has been updated to reflect what is in scope and what is not

**Review by astafan8 — APPROVED** [review]

File: qcodes/extensions/__init__.py [hunk 3]
```diff
@@ -0,0 +1,8 @@
+"""
+The extensions module contains smaller modules that extend the functionality of QCoDeS.
+These modules may import from all of QCoDeS but do not them self get imported into QCoDeS.
+"""
+from .installation import register_station_schema_with_vscode
+from .slack import Slack, SlackTimeoutWarning
```
[the remark this answers is not in the document]
Comment (jenshnielsen) [reply qcodes/extensions/__init__.py:6]: Yes, I figured that this should always have been a private method.

File: qcodes/utils/plotting.py [hunk 4]
```diff
@@ -253,6 +253,7 @@ def auto_color_scale_from_config(colorbar: matplotlib.colorbar.Colorbar,
             by the lower limit. Default value is read from
             ``config.plotting.auto_color_scale.color_under``.
     """
+    import qcodes
```
[the remark this answers is not in the document]
Comment (jenshnielsen) [reply qcodes/utils/plotting.py:256]: Yes but I have decided to leave the refactoring of utils to another pr.

File: qcodes/dataset/__init__.py [hunk 5]
```diff
@@ -20,7 +22,10 @@
     load_or_create_experiment,
     new_experiment,
 )
+from .experiment_settings import get_default_experiment_id, reset_default_experiment_id
+from .legacy_import import import_dat_file
 from .measurements import Measurement
+from .plotting import plot_by_id, plot_dataset
```
[the remark this answers is not in the document]
Comment (jenshnielsen) [reply qcodes/dataset/__init__.py:28]: For later I think

File: qcodes/utils/plotting.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (jenshnielsen) [reply qcodes/utils/plotting.py:256]: IMHO all this code should probably go into the dataset module (and not be public)

File: qcodes/utils/dataset/doNd.py [hunk 6]
```diff
@@ -1,1054 +1,31 @@
...
-import os
-import sys
-import time
-from abc import ABC, abstractmethod
-from contextlib import ExitStack, contextmanager
-from typing import Callable, Dict, Iterator, List, Optional, Sequence, Tuple, Union
-
-import matplotlib
-import numpy as np
-from tqdm.auto import tqdm
```
Comment (jenshnielsen) [inline qcodes/utils/dataset/doNd.py:11]: Need to review if some of the type definitions etc need to be reexported here

**Comment by jenshnielsen** [comment]
bors merge

**Pull request merged**